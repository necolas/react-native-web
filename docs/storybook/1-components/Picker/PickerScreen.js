/* eslint-disable react/jsx-sort-props */

/**
 * @flow
 */

import React from 'react';
import PropOnValueChange from './examples/PropOnValueChange';
import PropSelectedValue from './examples/PropSelectedValue';
import UIExplorer, { AppText, Description, DocItem, Section, storiesOf } from '../../ui-explorer';

const PickerScreen = () => (
  <UIExplorer title="Picker" url="1-components/Picker">
    <Description>
      <AppText>Renders the native &lt;select&gt; component.</AppText>
    </Description>
    <Section title="Props">
      <DocItem
        name="onValueChange"
        typeInfo="?function"
        description="Callback for when an item is selected. This is called with the following parameters: - itemValue: the value prop of the item that was selected - itemPosition: the index of the selected item in this picker."
        example={{
          render: () => <PropOnValueChange />
        }}
      />

      <DocItem
        name="selectedValue"
        typeInfo="?string"
        description="Value matching value of one of the items. Can be a string or an integer."
        example={{
          render: () => <PropSelectedValue />
        }}
      />

      <DocItem name="style" typeInfo="?pickerStyleType" />

      <DocItem
        name="testID"
        typeInfo="?string"
        description="Used to locate this view in end-to-end tests."
      />
    </Section>
  </UIExplorer>
);

storiesOf('Components', module).add('Picker', PickerScreen);
