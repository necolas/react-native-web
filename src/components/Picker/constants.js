export const MODE_DIALOG = MODE_DIALOG;
export const MODE_DROPDOWN = MODE_DROPDOWN;
