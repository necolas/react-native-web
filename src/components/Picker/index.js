/**
 * Copyright (c) 2017-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @providesModule Picker
 * @flow
 */

import applyNativeMethods from '../../modules/applyNativeMethods';
import { Component } from 'react';
import ColorPropType from '../../propTypes/ColorPropType';
import createElement from '../../modules/createElement';
import PickerItemPropType from './PickerItemPropType';
import PickerStylePropType from './PickerStylePropType';
import StyleSheetPropType from '../../propTypes/StyleSheetPropType';
import StyleSheet from '../../apis/StyleSheet';
import { arrayOf, func, number, oneOfType, string } from 'prop-types';

class Picker extends Component {
  static propTypes = {
    children: arrayOf(PickerItemPropType),
    onValueChange: func,
    selectedValue: oneOfType([number, string]),
    style: StyleSheetPropType(PickerStylePropType),
    testID: string
  };

  static Item;

  render() {
    const { children, selectedValue, style, testID } = this.props;
    const props = {
      children,
      value: selectedValue,
      onChange: this._onChange,
      style: [styles.initial, style],
      testID
    };
    return createElement('select', props);
  }

  _onChange = (e: Object) => {
    const { selectedIndex, value } = e.target;
    this.props.onValueChange(value, selectedIndex);
  };
}

class PickerItem extends Component {
  static propTypes = {
    color: ColorPropType,
    label: string.isRequired,
    testID: string,
    value: oneOfType([number, string])
  };

  render() {
    const { label, value } = this.props;
    const props = {
      label,
      value
    };
    return createElement('option', props);
  }
}

const styles = StyleSheet.create({
  initial: {
    fontFamily: 'inherit',
    fontSize: 'inherit',
    margin: 0
  }
});

Picker.Item = PickerItem;

export default applyNativeMethods(Picker);
