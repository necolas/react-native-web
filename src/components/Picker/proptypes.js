import React from 'react';
import ColorPropType from '../../propTypes/ColorPropType';
import StyleSheetPropType from '../../propTypes/StyleSheetPropType';
import TextStylePropTypes from '../Text/TextStylePropTypes';
import ViewStylePropTypes from '../View/ViewStylePropTypes';
import View from '../View';
import {
  MODE_DIALOG,
  MODE_DROPDOWN
} from './constants';

const itemStylePropType = StyleSheetPropType(TextStylePropTypes);

const pickerStyleType = StyleSheetPropType({
  ...ViewStylePropTypes,
  color: ColorPropType
});

export const PICKER_PROPTYPES = {
  ...View.propTypes,
  /**
   * If set to false, the picker will be disabled, i.e. the user will not be able to make a
   * selection.
   */
  enabled: React.PropTypes.bool,
  /**
   * Style to apply to each of the item labels.
   */
  itemStyle: itemStylePropType,
  /**
   * On Android, specifies how to display the selection items when the user taps on the picker:
   *
   *   - 'dialog': Show a modal dialog. This is the default.
   *   - 'dropdown': Shows a dropdown anchored to the picker view
   *
   */
  mode: React.PropTypes.oneOf([ MODE_DIALOG, MODE_DROPDOWN ]),
  /**
   * Callback for when an item is selected. This is called with the following parameters:
   *   - `itemValue`: the `value` prop of the item that was selected
   *   - `itemPosition`: the index of the selected item in this picker
   */
  onValueChange: React.PropTypes.func,
  /**
   * Prompt string for this picker
   */
  prompt: React.PropTypes.string,
  /**
   * Value matching value of one of the items. Can be a string or an integer.
   */
  selectedValue: React.PropTypes.any,
  style: pickerStyleType,
  /**
   * Used to locate this view in end-to-end tests.
   */
  testID: React.PropTypes.string
};

export const PICKER_ITEM_PROPTYPES = {
  /**
   * Color of this item's text.
   */
  color: ColorPropType,
  /**
   * Text to display for this item.
   */
  label: React.PropTypes.string.isRequired,
  /**
   * Used to locate the item in end-to-end tests.
   */
  testID: React.PropTypes.string,
  /**
   * The value to be passed to picker's `onValueChange` callback when
   * this item is selected. Can be a string or an integer.
   */
  value: React.PropTypes.any
};
