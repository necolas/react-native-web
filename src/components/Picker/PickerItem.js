import React from 'react';
import { PICKER_ITEM_PROPTYPES } from './proptypes';

export default class PickerItem extends React.Component {

  static propTypes = PICKER_ITEM_PROPTYPES;

  render() {
    return null;
  }
}
