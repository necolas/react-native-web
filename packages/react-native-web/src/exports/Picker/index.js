/**
 * Copyright (c) 2017-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @providesModule Picker
 * @flow
 */

import applyNativeMethods from '../../modules/applyNativeMethods';
import { Component } from 'react';
import createElement from '../createElement';
import PickerItem from './PickerItem';
import PickerItemPropType from './PickerItemPropType';
import PickerStylePropTypes from './PickerStylePropTypes';
import StyleSheetPropType from '../../modules/StyleSheetPropType';
import StyleSheet from '../StyleSheet';
import TextPropTypes from '../Text/TextPropTypes';
import { arrayOf, bool, func, number, oneOfType, string } from 'prop-types';

const pickerStyleType = StyleSheetPropType(PickerStylePropTypes);

type Props = {
  children?: Array<typeof PickerItem>,
  enabled?: boolean,
  onValueChange?: Function,
  selectedValue?: number | string,
  style?: pickerStyleType,
  testID?: string,
  /* compat */
  itemStyle?: TextPropTypes.style,
  mode?: string,
  prompt?: string
};

class Picker extends Component<Props> {
  static propTypes = {
    children: arrayOf(PickerItemPropType),
    enabled: bool,
    onValueChange: func,
    selectedValue: oneOfType([number, string]),
    style: pickerStyleType,
    testID: string
  };

  static Item = PickerItem;

  render() {
    const {
      children,
      enabled,
      selectedValue,
      style,
      testID,
      /* eslint-disable */
      itemStyle,
      mode,
      prompt
      /* eslint-enable */
    } = this.props;

    return createElement('select', {
      children,
      disabled: enabled === false ? true : undefined,
      onChange: this._handleChange,
      style: [styles.initial, style],
      testID,
      value: selectedValue
    });
  }

  _handleChange = (e: Object) => {
    const { onValueChange } = this.props;
    const { selectedIndex, value } = e.target;
    if (onValueChange) {
      onValueChange(value, selectedIndex);
    }
  };
}

const styles = StyleSheet.create({
  initial: {
    fontFamily: 'System',
    fontSize: 'inherit',
    margin: 0
  }
});

export default applyNativeMethods(Picker);
