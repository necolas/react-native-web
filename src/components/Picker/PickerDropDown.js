import React, { Component } from 'react';
import PickerItem from './PickerItem';
import { PICKER_PROPTYPES } from './proptypes';

class PickerDropDown extends Component {
  static propTypes = PICKER_PROPTYPES;

  childHasColor() {
    for (const item in this.props.children) {
      if (item && item.props && item.props.color) {
        return true;
      }
    }
    return false;
  }

  render() {
    const {
      children = [],
      itemStyle,
      testID,
      style,
      selectedValue = '',
      onValueChange,
      enabled = true,
      prompt,
      ...rest
    } = this.props;
    if (this.childHasColor()) {
      return null;
    }
    const change = {};
    if (onValueChange) {
      change.onChange = (e) => onValueChange(e.target.value, e.target.selectedIndex);
    }
    return (
      <select
        {...rest}
        {...change}
        data-testID={testID}
        disabled={!enabled}
        style={style}
        title={prompt}
        value={selectedValue}
      >
        {children.map((item, index) => {
          return (
            <PickerDropDown.Item
              key={`${testID}${index}`}
              {...item.props}
              {...itemStyle}
            />);
        })}
      </select>
    );
  }
}

PickerDropDown.Item = class extends PickerItem {
  render() {
    const { label, value, color, testID, ...rest } = this.props;
    if (color) {
      return null;
    }
    return (
      <option {...rest} data-testID={testID} value={ value } >
        { label }
      </option>
    );
  }
};

export default PickerDropDown;
