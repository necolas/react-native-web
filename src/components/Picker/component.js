import React, { Component } from 'react';
import PickerDropDown from './PickerDropDown';
import PickerDialog from './PickerDialog';
import PickerItem from './PickerItem';
import { PICKER_PROPTYPES } from './proptypes';
import {
  MODE_DIALOG,
  MODE_DROPDOWN
} from './constants';

class Picker extends Component {
  static MODE_DIALOG = MODE_DIALOG;

  static MODE_DROPDOWN = MODE_DROPDOWN;

  static defaultProps = {
    mode: MODE_DIALOG
  };

  static propTypes = PICKER_PROPTYPES;

  render() {
    const { mode } = this.props;
    if (mode === MODE_DIALOG) {
      return (<PickerDialog {...this.props} >{this.props.children}</PickerDialog>);
    } else if (mode === MODE_DROPDOWN) {
      return (<PickerDropDown {...this.props} >{this.props.children}</PickerDropDown>);
    }
    return null;
  }
}

/**
 * Individual selectable item in a Picker.
 */
Picker.Item = PickerItem;

export default Picker;

