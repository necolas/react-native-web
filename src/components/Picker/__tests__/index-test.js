/* eslint-env jasmine, jest */

import React from 'react';
import { shallow } from 'enzyme';
import Picker from '..';

describe('components/Picker', () => {
  describe('children', () => {
    test('when contains options, renders a select element with options', () => {
      const pickerWithItems = (
        <Picker>
          <Picker.Item label="Superman" value="1" />
          <Picker.Item label="Wonder Woman" value="2" />
        </Picker>
      );
      const component = shallow(pickerWithItems);
      expect(component).toMatchSnapshot();
    });
  });

  describe('onValueChange', () => {
    test('when changed, event is fired with (value, index)', () => {
      const onValueChange = jest.fn();
      const picker = (
        <Picker onValueChange={onValueChange} selectedValue="1">
          <Picker.Item label="Superman" value="1" />
          <Picker.Item label="Wonder Woman" value="2" />
        </Picker>
      );
      const component = shallow(picker);
      component.find('select').simulate('change', {
        target: { selectedIndex: '1', value: '2' }
      });
      expect(onValueChange).toHaveBeenCalledWith('2', '1');
    });
  });

  describe('selectedValue', () => {
    test('when "2":string, item is selected', () => {
      const pickerWithSelection = (
        <Picker selectedValue="2">
          <Picker.Item label="Superman" value="1" />
          <Picker.Item label="Wonder Woman" value="2" />
        </Picker>
      );
      const component = shallow(pickerWithSelection);
      expect(component.find('select').prop('value')).toBe('2');
    });
    test('when 2:number, item is selected', () => {
      const pickerWithNumberSelection = (
        <Picker selectedValue={2}>
          <Picker.Item label="Superman" value={1} />
          <Picker.Item label="Wonder Woman" value={2} />
        </Picker>
      );
      const component = shallow(pickerWithNumberSelection);
      expect(component.find('select').prop('value')).toBe(2);
    });
  });

  describe('onValueChange', () => {
    test('when changed, event is fired with (value, index)', () => {
      const onValueChange = jest.fn();
      const picker = (
        <Picker onValueChange={onValueChange} selectedValue="1">
          <Picker.Item label="Superman" value="1" />
          <Picker.Item label="Wonder Woman" value="2" />
        </Picker>
      );
      const component = shallow(picker);
      component.find('select').simulate('change', {
        target: { selectedIndex: '1', value: '2' }
      });
      expect(onValueChange).toHaveBeenCalledWith('2', '1');
    });
  });
});
