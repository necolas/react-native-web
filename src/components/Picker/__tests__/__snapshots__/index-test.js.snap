// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Picker children when contains options, renders a select element with options 1`] = `
<select
  className="rn-fontFamily-poiln3 rn-fontSize-7cikom rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw"
  onChange={[Function]}
>
  <PickerItem
    label="Superman"
    value="1"
  />
  <PickerItem
    label="Wonder Woman"
    value="2"
  />
</select>
`;
